# Intro to Linear Regression with Scikit-learn

Full tutorial [here](https://github.com/justmarkham/scikit-learn-videos/blob/master/06_linear_regression.ipynb)

In [1]:
import pandas as pd
data = pd.read_csv('http://www-bcf.usc.edu/~gareth/ISL/Advertising.csv', index_col=0)
data.head()

TV  Radio  Newspaper  Sales
1  230.1   37.8       69.2   22.1
2   44.5   39.3       45.1   10.4
3   17.2   45.9       69.3    9.3
4  151.5   41.3       58.5   18.5
5  180.8   10.8       58.4   12.9

In [4]:
# Create feature DataFrame
features = data[['TV', 'Radio', 'Newspaper']]
features.head()

TV  Radio  Newspaper
1  230.1   37.8       69.2
2   44.5   39.3       45.1
3   17.2   45.9       69.3
4  151.5   41.3       58.5
5  180.8   10.8       58.4

In [5]:
# Create label DataFrame
labels = data[['Sales']]
labels.head()

Sales
1   22.1
2   10.4
3    9.3
4   18.5
5   12.9

### Split into Training and Test datasets

In [23]:
from sklearn.cross_validation import train_test_split

features_train, features_test, labels_train, labels_test = train_test_split(features, labels, random_state=1)

# default split is 75% for training and 25% for testing
print(features_train.shape)
print(labels_train.shape)
print(features_test.shape)
print(labels_test.shape)

(150, 3)
(150, 1)
(50, 3)
(50, 1)


### Linear Regression Model

In [24]:
from sklearn.linear_model import LinearRegression

linreg = LinearRegression()

# fit the model to the training data (learn the coefficients)
linreg.fit(features_train, labels_train)

LinearRegression(copy_X=True, fit_intercept=True, n_jobs=1, normalize=False)

### Intercepts and Coefficients

![Formula](https://render.githubusercontent.com/render/math?math=y%20%3D%20%5Cbeta_0%20%2B%20%5Cbeta_1x_1%20%2B%20%5Cbeta_2x_2%20%2B%20...%20%2B%20%5Cbeta_nx_n&mode=inline)

* $y$ is the label (predicted value)
* $\beta_0$ is the intercept
* $\beta_1$ is the coefficient for $x_1$ (the first feature)
* $\beta_n$ is the coefficient for $x_n$ (the nth feature)

In [30]:
# Intercept and Coefficients
print linreg.intercept_
print linreg.coef_

# pair the feature names with the coefficients
feature_names = ['TV', 'Radio', 'Newspaper']
print zip(feature_names, linreg.coef_[0])

[ 2.87696662]
[[ 0.04656457  0.17915812  0.00345046]]
[('TV', 0.046564567874150288), ('Radio', 0.17915812245088836), ('Newspaper', 0.0034504647111803788)]


**How do we interpret the TV coefficient (0.0466)?**

For every 1 TV, sales increase .0466 units

In [38]:
#Sales Prediction
sales_predictions = linreg.predict(features_test)
sales_predictions[0:3]   #numpy array

array([[ 21.70910292],
       [ 16.41055243],
       [  7.60955058]])

### Model evaluation metrics for regression

**Mean Absolute Error (MAE)** is the mean of the absolute value of the errors:

In [45]:
from sklearn import metrics

# calculate MAE by hand
(10 + 0 + 20 + 10) / 4.

metrics.mean_absolute_error(labels_test, sales_predictions)

1.0668917082595208

**Mean Squared Error (MSE)** is the mean of the squared errors:

In [47]:
#calculate MSE by hand
(10**2 + 0**2 + 20**2 + 10**2) / 4.

metrics.mean_squared_error(labels_test, sales_predictions)

1.9730456202283375

**Root Mean Squared Error (RMSE)** is the square root of the mean of the squared errors

In [49]:
# calculate RMSE by hand
import numpy as np
np.sqrt((10**2 + 0**2 + 20**2 + 10**2)/4.)

# calculate RMSE using scikit-learn
np.sqrt(metrics.mean_squared_error(labels_test, sales_predictions))

1.4046514230328953

In [51]:
#Let's use RMSE b/c it's interpretable in 'label' units
sales_predictions = np.sqrt(metrics.mean_squared_error(labels_test, sales_predictions))
sales_predictions

1.4046514230328953

### Feature selection

In [60]:
# Let's remove 'Newspaper' and see if we can reduce the RMSE
features_to_include = ['TV','Radio']

# use the list to select a subset of the original DataFrame
X = data[features_to_include]

# select a Series from the DataFrame
y = data.Sales

# split into training and testing sets
X_train, X_test, y_train, y_test = train_test_split(X, y, random_state=1)

# fit the model to the training data (learn the coefficients)
linreg.fit(X_train, y_train)

# make predictions on the testing set
y_pred = linreg.predict(X_test)

# compute the RMSE of our predictions
np.sqrt(metrics.mean_squared_error(y_test, y_pred))

1.3879034699382886

It worked! A lower number for RMSE is better.